tests/data/test_data_pipeline.py

```diff
@@ -0,0 +1,750 @@
+    assert _seq.per_sample_pre_tensor_transform.func == preprocess.per_sample_pre_tensor_transform
+    assert _seq.per_sample_to_tensor_transform.func == preprocess.per_sample_to_tensor_transform
+    assert _seq.per_sample_post_tensor_transform.func == preprocess.train_per_sample_post_tensor_transform
+    assert train_worker_preprocessor.collate_fn.func == default_collate
+    assert train_worker_preprocessor.per_batch_transform.func == preprocess.per_batch_transform
+
+    _seq = val_worker_preprocessor.per_sample_transform
+    assert _seq.per_sample_pre_tensor_transform.func == preprocess.val_per_sample_pre_tensor_transform
+    assert _seq.per_sample_to_tensor_transform.func == preprocess.per_sample_to_tensor_transform
+    assert _seq.per_sample_post_tensor_transform.func == preprocess.per_sample_post_tensor_transform
+    assert val_worker_preprocessor.collate_fn.func == data_pipeline._do_nothing_collate
+    assert val_worker_preprocessor.per_batch_transform.func == preprocess.per_batch_transform
+
+    _seq = test_worker_preprocessor.per_sample_transform
+    assert _seq.per_sample_pre_tensor_transform.func == preprocess.per_sample_pre_tensor_transform
+    assert _seq.per_sample_to_tensor_transform.func == preprocess.per_sample_to_tensor_transform
+    assert _seq.per_sample_post_tensor_transform.func == preprocess.per_sample_post_tensor_transform
+    assert test_worker_preprocessor.collate_fn.func == preprocess.test_collate
+    assert test_worker_preprocessor.per_batch_transform.func == preprocess.per_batch_transform
+
+    _seq = predict_worker_preprocessor.per_sample_transform
+    assert _seq.per_sample_pre_tensor_transform.func == preprocess.per_sample_pre_tensor_transform
+    assert _seq.per_sample_to_tensor_transform.func == preprocess.predict_per_sample_to_tensor_transform
+    assert _seq.per_sample_post_tensor_transform.func == preprocess.per_sample_post_tensor_transform
+    assert predict_worker_preprocessor.collate_fn.func == default_collate
+    assert predict_worker_preprocessor.per_batch_transform.func == preprocess.per_batch_transform
+
+
+class CustomPreprocess(Preprocess):
+
+    def train_per_sample_transform(self, *_, **__):
+        pass
+
+    def train_per_batch_transform_on_device(self, *_, **__):
+        pass
+
+    def test_per_sample_transform(self, *_, **__):
+        pass
+
+    def test_per_batch_transform(self, *_, **__):
+        pass
+
+    def test_per_sample_transform_on_device(self, *_, **__):
+        pass
+
+    def test_per_batch_transform_on_device(self, *_, **__):
+        pass
+
+    def val_per_batch_transform(self, *_, **__):
+        pass
+
+    def val_per_sample_transform_on_device(self, *_, **__):
+        pass
+
+    def predict_per_sample_transform(self, *_, **__):
+        pass
+
+    def predict_per_sample_transform_on_device(self, *_, **__):
+        pass
+
+    def predict_per_batch_transform_on_device(self, *_, **__):
+        pass
```

Comment:
all are returning none, so what does it mean?

Reply:
They are not called. Just checking that the function are properly gathered for each stage.